source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,489 @@
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => {
+                tracing::trace!("CMD8 response: {value:#x}");
+                if value == 0x1AA {
+                    card_type = CardType::SD2;
+                }
+            }
+            Response::Long(_) => return Err(Error::Response),
+        }
+
+        // TODO: limit the number of attempts
+        let ocr = loop {
+            // Go to *APP* mode before sending application command
+            let _ = self
+                .cmd(Command {
+                    index: 55,
+                    argument: 0,
+                    options: HardwareOptions::None,
+                    kind: CommandKind::Control,
+                    rsp_type: ResponseType::Short,
+                    rsp_crc: true,
+                    buffer: None,
+                })
+                .await?;
+
+            let mut op_cond_arg = OCR_VOLTAGE_MASK & 0x00ff8000;
+            if card_type != CardType::SD1 {
+                op_cond_arg |= OCR_HCS;
+            }
+            match self
+                .cmd(Command {
+                    index: 41,
+                    argument: op_cond_arg,
+                    options: HardwareOptions::None,
+                    kind: CommandKind::Control,
+                    rsp_type: ResponseType::Short,
+                    rsp_crc: false,
+                    buffer: None,
+                })
+                .await?
+            {
+                Response::Short { value, .. } => {
+                    tracing::trace!("ACMD41 response: {value:#x}");
+                    if value & OCR_NBUSY == OCR_NBUSY {
+                        // Card has finished power up, data is valid
+                        break value;
+                    }
+                }
+                Response::Long(_) => return Err(Error::Response),
+            }
+
+            // TODO: wait some time (e.g., 1ms?)
+        };
+
+        if (ocr & OCR_HCS) == OCR_HCS {
+            card_type = CardType::SDHC;
+        }
+
+        Ok(card_type)
+    }
+
+    /// Get the card identification register
+    pub async fn get_cid(&mut self) -> Result<CardIdentification, Error> {
+        match self
+            .cmd(Command {
+                index: 2,
+                argument: 0,
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Long,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { .. } => Err(Error::Response),
+            Response::Long(value) => {
+                tracing::trace!("CMD2 response: {value:?}");
+                union CidResponse {
+                    value: [u32; 4],
+                    cid: u128,
+                }
+                let rsp = CidResponse { value };
+                Ok(CardIdentification(unsafe { rsp.cid }))
```

Comment:
Looking at this again, it indeed makes sense to let the hardware specific driver do the conversion.
This is currently `[u32; 4]` because in the D1 you need to retrieve 4 32-bit registers to reconstruct the "long" response, but this shouldn't be part of the service interface.

Reply:
yeah, i think that's the right thing --- from the perspective of the hardware-agnostic `SdmmcService`, it's just a 128-bit number, i think. how that's constructed from register values or whatever is up to the driver implementation.